### ISSUE TITLE
Inspect the Expires header before updating feed

### ISSUE THREAD

**Issue by SeanBE** [newsboat/newsboat#1138]
I've run into issues with the RSS feeds provided by https://arxiv.org. These feeds are updated daily somewhere between 00:30 GMT and 01:50 GMT *.

* If you run a periodic refresh for newsboat within arxiv's update window, you will never receive new items (after initial load)
* They update the Last-Modified and Etag headers with an empty list first before actually populating the list later. (I'm assuming some sort of batch job)
* They also set the header Expires to -1 during this window. 

The question is should we inspect the `Expires` header before looking at the other cache headers and updating the RSS feed?

Some extra reading material:

Logs during this time
```
[2020-08-11 00:54:28] DEBUG: RssParser::download_http: lm = 1597105800 etag = "Tue, 11 Aug 2020 00:30:00 GMT", "1597105800"
[2020-08-11 00:54:28] DEBUG: RssParser::download_http: lastmodified old: 1597019400 new: 1597105800
[2020-08-11 00:54:28] DEBUG: RssParser::download_http: etag old: "Mon, 10 Aug 2020 00:30:00 GMT" new "Tue, 11 Aug 2020 00:30:00 GMT", "1597105800"
[2020-08-11 00:54:28] DEBUG: running query: UPDATE rss_feed SET lastmodified = '1597105800', etag = '"Tue, 11 Aug 2020 00:30:00 GMT", "1597105800"' WHERE rssurl = 'http://arxiv.org/rss/q-fin/new'
[2020-08-11 00:54:28] DEBUG: RssParser::parse: http URL http://arxiv.org/rss/q-fin/new, valid: true
[2020-08-11 00:54:28] DEBUG: RssParser::parse: feed title = `q-fin updates on arXiv.org' link = `http://arxiv.org/'
[2020-08-11 00:54:28] DEBUG: Cache::remove_old_deleted_items: not cleaning up anything because last reload brought no new items (detected no changes)
[2020-08-11 00:54:28] DEBUG: ScopeMeasure: function `Cache::remove_old_deleted_items' took 0.000010 s
[2020-08-11 00:54:28] DEBUG: Controller::replace_feed: saving
[2020-08-11 00:54:28] DEBUG: running query: SELECT count(*) FROM rss_feed WHERE rssurl = 'http://arxiv.org/rss/q-fin/new';
[2020-08-11 00:54:28] DEBUG: Cache::externalize_rss_feed: rss_feeds with rssurl = 'http://arxiv.org/rss/q-fin/new': found 1
[2020-08-11 00:54:28] DEBUG: running query: UPDATE rss_feed SET title = 'q-fin updates on arXiv.org', url = 'http://arxiv.org/', is_rtl = 0 WHERE rssurl = 'http://arxiv.org/rss/q-fin/new';
```

An example of the arxiv response during this window
```
HTTP/1.1 200 OK
Date: Tue, 11 Aug 2020 00:54:14 GMT
Server: Apache
ETag: "Tue, 11 Aug 2020 00:30:00 GMT", "1597105800"
Expires: -1
Last-Modified: Tue, 11 Aug 2020 00:30:00 GMT
Vary: Accept-Encoding,User-Agent
Transfer-Encoding: chunked
Content-Type: text/xml

<?xml version="1.0" encoding="UTF-8"?>

<rdf:RDF
...
>

<channel rdf:about="http://arxiv.org/">
<title>cs updates on arXiv.org</title>
<link>http://arxiv.org/</link>
<description rdf:parseType="Literal">Computer Science (cs) updates on the arXiv.org e-print archive</description>
<dc:language>en-us</dc:language>
<dc:date>2020-08-10T20:30:00-05:00</dc:date>
<dc:publisher>www-admin@arxiv.org</dc:publisher>
<dc:subject>Computer Science</dc:subject>
<syn:updateBase>1901-01-01T00:00+00:00</syn:updateBase>
<syn:updateFrequency>1</syn:updateFrequency>
<syn:updatePeriod>daily</syn:updatePeriod>
<items>
 <rdf:Seq>
 </rdf:Seq>
</items>
</rdf:RDF>
```

Headers from a successful call an hour later:
```
HTTP/1.1 200 OK
Date: Tue, 11 Aug 2020 01:56:14 GMT
Server: Apache
ETag: "Tue, 11 Aug 2020 00:30:00 GMT", "1597105800"
Expires: Wed, 12 Aug 2020 00:00:00 GMT
Last-Modified: Tue, 11 Aug 2020 00:30:00 GMT
Vary: Accept-Encoding,User-Agent
Transfer-Encoding: chunked
Content-Type: text/xml
```

**Comment by Minoru** [comment]
> The question is should we inspect the Expires header before looking at the other cache headers and updating the RSS feed?

Most likely. We can also inspect `Cache-Control` header; its `max-age` directive is essentially the same as `Expires`. All that stuff is described in [RFC 7234](https://tools.ietf.org/html/rfc7234).

I foresee a couple problems with those headers, though:

1. if a misconfigured server sends a date that's far in the future, Newsboat will effectively stop updating that feed.

    This isn't a novel problem. `ETag` and `Last-Modified` are susceptible to it as well. We already have an `always-download` setting to work around that. The only real risk here is that by examining more headers, we increase the chance of running into a broken one. Then the user will have to debug it (like you did!) and resort to `always-download`;

2. before making a request, Newsboat will compare the "expiry time" (the last seen value of the `Expires` header) to the local time. If local clock is misconfigured, Newsboat will either always make a request, or never make a request.

    In the first case, `ETag` and `Last-Modified` can still save the bandwidth and time, so that's okay. In the second case, though, the feed will stop updating, making the user upset again.

I don't think these can be solved in general, but we can mitigate them by copying what browsers do: provide a `force-reload` operation that reloads a feed ignoring `Expires`, `ETag` and `Last-Modified`. That again makes the user unhappy, but perhaps makes it slightly easier to debug. We can also arbitrarily cap the "expiry date" at e.g. one day into the future, lessening the problem.

<hr/>

But none of that will help with arXiv. The `Expires` header specifies the date *before which* the client should just use a cached response. Once that date passes, the client can validate their cache, and [RFC 7234 §4.3](https://tools.ietf.org/html/rfc7234#section-4.3) actually suggests conditional requests for that, i.e. using `ETag` and `Last-Modified` values. If we ignore those two headers, then arXiv's `Expires` header starts to make sense: it helps "regulars" get their updates right after midnight and then asks them to stay silent for a day, freeing bandwidth for "random newcomers" who fetch the feed for the first time.

Now, the `ETag` that arXiv provides doesn't conform to the standard: [RFC 7232 §2.3](https://tools.ietf.org/html/rfc7232#section-2.3) mandates that the ETag should start and end with a double quotes, and *there shouldn't be double quotes in between*. arXiv's `"Tue, 11 Aug 2020 00:30:00 GMT", "1597105800"` obviously breaks that. The RFC doesn't say what to do with malformed headers. Newsboat accepts the malformed value and sends it back with `If-None-Match`.

But that's what we have to work with, so let's proceed. Since the value doesn't start with `W/`, we assume it's a strong validator ([that's the default](https://tools.ietf.org/html/rfc7232#section-2.3)). Quoting RFC 7232 §2.1:

> A "strong validator" is representation metadata that changes value whenever a change occurs to the representation data that would be observable in the payload body of a 200 (OK) response to GET.

If I'm reading this correctly, this means arXiv *must* change the ETag when they fill out the feed, because the feed is clearly observable in the payload.

As for `Last-Modified`, it's irrelevant, since [RFC 7232 §3.3](https://tools.ietf.org/html/rfc7232#section-3.3) says that the server *must* ignore the `If-Modified-Since` (i.e. the date) if `If-None-Match` (i.e. the ETag) is present. If it were relevant, it would have been a strong validator for 99.93% of the time (1439 minutes out of 1440 in a day) according to the second rule in [RFC 7232 §2.2.2](https://tools.ietf.org/html/rfc7232#section-2.2.2) and considering that arXiv updates the feed once a day. This again means that arXiv *must* change Last-Modified when they fill out the feed.

In short: I believe arXiv implements `ETag` and `Last-Modified` response headers incorrectly, and their `Expires` header only helps those who ignore the other two headers. Your only option is to add their feed(s) to `always-download`. I'll shoot them an email as we discussed on IRC, and let's hope they have the time to actually fix this.

**Comment by Minoru** [comment]
> I'll shoot them an email as we discussed on IRC, and let's hope they have the time to actually fix this.

The info I provided was passed on to their programmers. Hopefully this will get fixed :crossed_fingers:

**Comment by SeanBE** [comment]
Looks like they finally did it! Follows 2.0 spec and returns some better caching headers.
https://blog.arxiv.org/2024/01/31/attention-arxiv-users-re-implemented-rss/

```
HTTP/2 200
content-type: application/rss+xml
etag: 8d7b3c3a61616dc40dc6dee5f47bc8d431f4fb0eedc0271619fa04be3af60fda
cache-control: max-age=86393
x-cloud-trace-context: 150f2acae7f99ca57b93a5b3dcbd4df4
server: Google Frontend
via: 1.1 google, 1.1 google, 1.1 varnish, 1.1 varnish
accept-ranges: bytes
date: Tue, 13 Feb 2024 05:41:56 GMT
age: 2510
x-served-by: cache-lga21955-LGA, cache-lcy-eglc8600040-LCY
x-cache: MISS, HIT
x-cache-hits: 0, 1
x-timer: S1707802916.448162,VS0,VE3
strict-transport-security: max-age=300
content-length: 2241766
```
